# **S&P 500 Analysis**
*By Justin Kim*  
[GitHub Repository](https://github.com/iluvcoding123/sp500_analysis)


## Introduction
The S&P 500 is a stock market index comprised of the 500 largest publicly traded companies in the U.S. Netflix, Amazon, Coca-Cola, and Apple are but a few notable names among its top performers. It serves as a key benchmark for the performance of the U.S. stock market and overall economy. By tracking the performance of 500 leading companies, it provides valuable insights into economic trends and market conditions. In this project, we will aim to reveal such insights by using data analysis techniques in Python.

## Data Overview
We will be using three primary datasets sourced from [Kaggle](https://www.kaggle.com/datasets/andrewmvd/sp-500-stocks). 
1. `sp500_companies.csv`  
    Contains detailed information about each S&P 500 company including symbol, industry, market cap, weight, etc.

2. `sp500_index.csv`  
    Contains historical daily values of the S&P 500 index
    
3. `sp500_stocks.csv`  
    Contains historical daily open, close, and other related metrics for each company. 

### Importing Data
First, let's import the datasets into pandas DataFrames so we can analyze them throughout this project. Note: The `sp500_stocks.csv` file will be accessed externally on Google Drive due to its large file size (~200 MB).

In [8]:
import pandas as pd
import gdown
import os

# Download 'sp500_stocks.csv' from Google Drive
# (Stored externally due to large file size)
url = 'https://drive.google.com/uc?id=1ODTONeSfNwIa7ADrFSdTqiJHWHBiMV5M'
file_path = 'sp500_stocks.csv'
# Skip download if file already exists in path
if not os.path.exists(file_path):
    print("Downloading sp500_stocks.csv from Google Drive...")
    gdown.download(url, file_path, quiet=False)
else:
    print("sp500_stocks.csv download skipped.")

# Load the CSV files into DataFrames
companies = pd.read_csv('sp500_companies.csv')
index = pd.read_csv('sp500_index.csv')
stocks = pd.read_csv('sp500_stocks.csv')

sp500_stocks.csv download skipped.


## Exploratory Data Analysis
Now that our data is loaded into the `companies`, `index`, and `stocks` DataFrames, we can perform exploratory data analysis to gain a better understanding of the data's characteristics.

### 1. Data Overview and Structure

In [26]:
# Print a summary of each DataFrame's structure
print('companies info:')
print(companies.info())

print('\nindex info:')
print(index.info())

print('\nstocks info:')
print(stocks.info())

companies info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Exchange             503 non-null    object 
 1   Symbol               503 non-null    object 
 2   Shortname            503 non-null    object 
 3   Longname             503 non-null    object 
 4   Sector               503 non-null    object 
 5   Industry             503 non-null    object 
 6   Currentprice         503 non-null    float64
 7   Marketcap            503 non-null    int64  
 8   Ebitda               474 non-null    float64
 9   Revenuegrowth        501 non-null    float64
 10  City                 503 non-null    object 
 11  State                483 non-null    object 
 12  Country              503 non-null    object 
 13  Fulltimeemployees    498 non-null    float64
 14  Longbusinesssummary  503 non-null    object 
 15  Weight               503

### 2. Summary Statistics

In [25]:
# Print summary statistics for each DataFrame
print('companies summary statistics:')
print(companies.describe())

print('\nindex summary statistics:')
print(index.describe())

print('\nstocks summary statistics:')
print(stocks.describe())

companies summary statistics:
       Currentprice     Marketcap        Ebitda  Revenuegrowth  \
count    503.000000  5.030000e+02  4.740000e+02     501.000000   
mean     220.833926  1.039559e+11  6.799068e+09       0.057695   
std      507.631537  3.012797e+11  1.486008e+10       0.142051   
min        8.560000  6.334528e+09 -4.107000e+09      -0.573000   
25%       69.830000  1.917211e+10  1.585225e+09      -0.001000   
50%      123.800000  3.671813e+10  2.880500e+09       0.047000   
75%      234.515000  7.657803e+10  5.874562e+09       0.096000   
max     9425.370000  3.288951e+12  1.317810e+11       1.430000   

       Fulltimeemployees      Weight  
count       4.980000e+02  503.000000  
mean        5.815182e+04    0.001988  
std         1.390450e+05    0.005762  
min         2.800000e+01    0.000121  
25%         1.017825e+04    0.000367  
50%         2.137450e+04    0.000702  
75%         5.548300e+04    0.001464  
max         2.100000e+06    0.062899  

index summary statistic

### 3. Check for Missing Data

In [29]:
# Check for number of missing values in each DataFrame
print("companies missing values:")
print(companies.isnull().sum())

print("index missing values:")
print(index.isnull().sum())

print("stocks missing values:")
print(stocks.isnull().sum())

companies missing values:
Exchange                0
Symbol                  0
Shortname               0
Longname                0
Sector                  0
Industry                0
Currentprice            0
Marketcap               0
Ebitda                 29
Revenuegrowth           2
City                    0
State                  20
Country                 0
Fulltimeemployees       5
Longbusinesssummary     0
Weight                  0
dtype: int64
index missing values:
Date      0
S&P500    0
dtype: int64
stocks missing values:
Date             0
Symbol           0
Adj Close    94879
Close        94879
High         94879
Low          94879
Open         94879
Volume       94879
dtype: int64


Hmmm, seems like a non-negligible amount of data is missing from the `stocks` table. **94,879** values are missing from each of the `Adj Close`, `Close`, `High`, `Low`, `Open`, and `Volume` columns. Let's calculate exactly how much is missing relative to the total.

In [28]:
# Calculate and print the percentage of values that are missing for each column
missing_percentage = stocks.isnull().sum() / len(stocks) * 100
print(missing_percentage)

Date         0.000000
Symbol       0.000000
Adj Close    5.098007
Close        5.098007
High         5.098007
Low          5.098007
Open         5.098007
Volume       5.098007
dtype: float64


Interesting... we calculated a little over **5%** of values are missing from those columns. We'll have to keep that in mind going forward. For now, let's continue exploring the data. 